### PR TITLE
8257182: JCK test failures in integer / long rotation tests

### PR DESCRIPTION
**Pull request by chhagedorn**
The `RotateLeftNode::Value()` and `RotateRightNode::Value()` methods use an arithmetic instead of a logical shift due to applying the shift operator `>>` to a signed instead of an unsigned number. When, for example, rotating a negative even integer (least significant bit is 0) to the right, we should always get a positive number (most significant bit becoming 0). However, the arithmetic shift keeps the integer being negative. This lets some JCK rotation tests fail with -Xcomp.

The previously failing JCK tests work again with -Xcomp by applying this fix.

Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257182](https://bugs.openjdk.java.net/browse/JDK-8257182): JCK test failures in integer / long rotation tests

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1562/head:pull/1562`
`$ git checkout pull/1562`

### CODE REVIEW

File: src/hotspot/share/opto/mulnode.cpp [hunk 1]
```diff
@@ -1506,8 +1506,9 @@ const Type* RotateLeftNode::Value(PhaseGVN* phase) const {
       return r1;
     }
     if (r1->is_con() && r2->is_con()) {
-      int shift = r2->get_con() & (BitsPerJavaInteger - 1); // semantics of Java shifts
-      return TypeInt::make((r1->get_con() << shift) | (r1->get_con() >> (32 - shift)));
+      unsigned int r1_con = (unsigned int)r1->get_con();
+      unsigned int shift = (unsigned int)(r2->get_con()) & (unsigned int)(BitsPerJavaInteger - 1); // semantics of Java shifts
```
Comment (TheRealMDoerr) [inline src/hotspot/share/opto/mulnode.cpp:1510]: I'd prefer to use juint which actually is the same, but would be more consistent IMHO.

**Review by TheRealMDoerr — COMMENTED** [review]
Looks great! Thanks for fixing it and adding a regression test.

File: src/hotspot/share/opto/mulnode.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/mulnode.cpp:1510]: +1

Comment (chhagedorn) [reply src/hotspot/share/opto/mulnode.cpp:1510]: I agree, I'll change it to `juint`

**Comment by chhagedorn** [comment]
@TheRealMDoerr @vnkozlov Thanks for your reviews!

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by chhagedorn** [comment]
@iwanowww Thanks for your review!

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Review by TheRealMDoerr — APPROVED** [review]

**Comment by chhagedorn** [comment]
@TobiHartmann @vnkozlov @TheRealMDoerr Thanks for your reviews!

**Pull request closed without merging**